File: color_utils.py
from typing import Dict, List, Optional, Tuple
from PIL import Image
import math


def rgb_to_hex(r: int, g: int, b: int) -> str:
    return '#' + ''.join(f'{x:02x}' for x in [r, g, b])


def hex_to_rgb(hex_color: str) -> str:
    r = int(hex_color[1:3], 16)
    g = int(hex_color[3:5], 16)
    b = int(hex_color[5:7], 16)
    return f'rgb({r}, {g}, {b})'


def hex_to_rgba(hex_color: str, alpha: float = 1.0) -> str:
    r = int(hex_color[1:3], 16)
    g = int(hex_color[3:5], 16)
    b = int(hex_color[5:7], 16)
    return f'rgba({r}, {g}, {b}, {alpha})'


class ColorPalette:
    def __init__(self, hex_color: str, rgb: Tuple[int, int, int], count: int, percentage: int):
        self.hex = hex_color
        self.rgb = {'r': rgb[0], 'g': rgb[1], 'b': rgb[2]}
        self.count = count
        self.percentage = percentage


def extract_palette(image_path: str, options: Optional[Dict] = None) -> List[ColorPalette]:
    if options is None:
        options = {}
    
    max_colors = options.get('maxColors', 5)
    resolution = options.get('resolution', 100)
    exclude_bw = options.get('excludeBW', False)
    interpolation = options.get('interpolation', 'linear')
    dev = options.get('dev', False)
    
    interpolation_map = {
        'nearest': Image.NEAREST,
        'linear': Image.BILINEAR,
        'cubic': Image.BICUBIC,
        'lanczos': Image.LANCZOS
    }
    
    img = Image.open(image_path)
    original_width, original_height = img.size
    
    if dev:
        print('=== 调试信息 ===')
        print(f'图片路径: {image_path}')
        print(f'原始尺寸: {original_width} x {original_height}')
        print(f'分辨率参数: {resolution}')
        print(f'最大颜色数: {max_colors}')
        print(f'排除黑白: {exclude_bw}')
        print(f'插值模式: {interpolation}')
    
    if resolution == 1:
        width, height = original_width, original_height
        if dev:
            print(f'缩放模式: 保持原始尺寸 ({width} x {height})')
    elif resolution > 1:
        size = resolution
        if original_width > original_height:
            width = size
            height = int(round((original_height / original_width) * size))
        else:
            height = size
            width = int(round((original_width / original_height) * size))
        if dev:
            print(f'缩放模式: 固定尺寸模式 ({width} x {height})')
    else:
        scale = resolution
        width = max(1, int(round(original_width * scale)))
        height = max(1, int(round(original_height * scale)))
        if dev:
            print(f'缩放模式: 比例缩放模式 ({(scale * 100):.1f}%)')
            print(f'缩放后尺寸: {width} x {height}')
    
    resample = interpolation_map.get(interpolation, Image.BILINEAR)
    img = img.resize((width, height), resample)
    
    if img.mode != 'RGBA':
        img = img.convert('RGBA')
    
    pixels = list(img.getdata())
    
    if dev:
        print(f'处理后的尺寸: {width} x {height}')
        print(f'像素总数: {width * height}')
        print(f'数据长度: {len(pixels) * 4} bytes')
    
    color_counts: Dict[str, Dict] = {}
    total_valid_pixels = 0
    transparent_pixels = 0
    excluded_bw_pixels = 0
    bw_near_colors = set()
    
    factor = 12
    
    for r, g, b, a in pixels:
        if a < 128:
            transparent_pixels += 1
            continue
        
        qr = int(round(r / factor)) * factor
        qg = int(round(g / factor)) * factor
        qb = int(round(b / factor)) * factor
        key = f'{qr},{qg},{qb}'
        
        is_bw_near_color = (r + g + b < 60) or (r + g + b > 710) or (max(r, g, b) - min(r, g, b) < 15)
        
        if exclude_bw:
            if is_bw_near_color:
                excluded_bw_pixels += 1
                continue
        else:
            if is_bw_near_color:
                bw_near_colors.add(key)
        
        if key not in color_counts:
            color_counts[key] = {'r': qr, 'g': qg, 'b': qb, 'count': 0}
        color_counts[key]['count'] += 1
        total_valid_pixels += 1
    
    sorted_colors = sorted(color_counts.values(), key=lambda x: x['count'], reverse=True)
    
    if dev:
        print(f'\n=== 颜色统计 ===')
        print(f'总像素数: {width * height}')
        print(f'透明像素: {transparent_pixels}')
        print(f'排除的黑白像素: {excluded_bw_pixels}')
        print(f'有效像素: {total_valid_pixels}')
        print(f'量化后的颜色数量: {len(sorted_colors)}')
        print(f'\n=== 前10个颜色（按出现次数）===')
        for i, c in enumerate(sorted_colors[:10]):
            key = f"{c['r']},{c['g']},{c['b']}"
            is_bw_near = key in bw_near_colors
            bw_label = ' (黑白临近色)' if is_bw_near else ''
            print(f"{i + 1}. {rgb_to_hex(c['r'], c['g'], c['b'])} - 出现 {c['count']} 次 ({int(round((c['count'] / total_valid_pixels) * 100))}%){bw_label}")
    
    final_palette = []
    skipped_similar = 0
    
    for c in sorted_colors:
        if len(final_palette) >= max_colors:
            break
        
        is_too_similar = False
        for existing in final_palette:
            dr = existing.rgb['r'] - c['r']
            dg = existing.rgb['g'] - c['g']
            db = existing.rgb['b'] - c['b']
            distance = math.sqrt(dr * dr + dg * dg + db * db)
            if distance < 45:
                is_too_similar = True
                break
        
        if not is_too_similar:
            final_palette.append(ColorPalette(
                hex_color=rgb_to_hex(c['r'], c['g'], c['b']),
                rgb=(c['r'], c['g'], c['b']),
                count=c['count'],
                percentage=int(round((c['count'] / total_valid_pixels) * 100))
            ))
        else:
            skipped_similar += 1
    
    if dev:
        print(f'\n=== 最终结果 ===')
        print(f'跳过的相似颜色: {skipped_similar}')
        print(f'最终提取颜色数: {len(final_palette)}')
        for i, c in enumerate(final_palette):
            print(f"{i + 1}. {c.hex} - RGB({c.rgb['r']}, {c.rgb['g']}, {c.rgb['b']}) - {c.percentage}%")
        print('=== 调试结束 ===\n')
    
    if len(final_palette) == 0 and len(sorted_colors) > 0:
        for c in sorted_colors[:max_colors]:
            final_palette.append(ColorPalette(
                hex_color=rgb_to_hex(c['r'], c['g'], c['b']),
                rgb=(c['r'], c['g'], c['b']),
                count=c['count'],
                percentage=int(round((c['count'] / total_valid_pixels) * 100))
            ))
    
    return final_palette

File: colorUtils.js
const rgbToHex = (r, g, b) => {
  return '#' + [r, g, b].map(x => {
    const hex = x.toString(16);
    return hex.length === 1 ? '0' + hex : hex;
  }).join('');
};

const hexToRgb = (hex) => {
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  return `rgb(${r}, ${g}, ${b})`;
};

const hexToRgba = (hex, alpha = 1) => {
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  return `rgba(${r}, ${g}, ${b}, ${alpha})`;
};

const extractPalette = async (imagePath, options = {}) => {
  const {
    maxColors = 5,
    resolution = 100,
    excludeBW = false,
    interpolation = 'linear',
    dev = false
  } = options;

  const sharp = require('sharp');
  
  const metadata = await sharp(imagePath).metadata();
  const originalWidth = metadata.width;
  const originalHeight = metadata.height;
  
  if (dev) {
    console.log('=== 调试信息 ===');
    console.log(`图片路径: ${imagePath}`);
    console.log(`原始尺寸: ${originalWidth} x ${originalHeight}`);
    console.log(`分辨率参数: ${resolution}`);
    console.log(`最大颜色数: ${maxColors}`);
    console.log(`排除黑白: ${excludeBW}`);
    console.log(`插值模式: ${interpolation}`);
  }
  
  let width, height;
  
  if (resolution === 1) {
    width = originalWidth;
    height = originalHeight;
    if (dev) {
      console.log(`缩放模式: 保持原始尺寸 (${width} x ${height})`);
    }
  } else if (resolution > 1) {
    const size = resolution;
    if (originalWidth > originalHeight) {
      width = size;
      height = Math.round((originalHeight / originalWidth) * size);
    } else {
      height = size;
      width = Math.round((originalWidth / originalHeight) * size);
    }
    if (dev) {
      console.log(`缩放模式: 固定尺寸模式 (${width} x ${height})`);
    }
  } else {
    const scale = resolution;
    width = Math.max(1, Math.round(originalWidth * scale));
    height = Math.max(1, Math.round(originalHeight * scale));
    if (dev) {
      console.log(`缩放模式: 比例缩放模式 (${(scale * 100).toFixed(1)}%)`);
      console.log(`缩放后尺寸: ${width} x ${height}`);
    }
  }
  
  const { data, info } = await sharp(imagePath)
    .resize(width, height, { fit: 'fill', kernel: interpolation })
    .raw()
    .toBuffer({ resolveWithObject: true });

  if (dev) {
    console.log(`处理后的尺寸: ${info.width} x ${info.height}`);
    console.log(`像素总数: ${info.width * info.height}`);
    console.log(`数据长度: ${data.length} bytes`);
  }

  const colorCounts = {};
  let totalValidPixels = 0;
  let transparentPixels = 0;
  let excludedBWPixels = 0;
  const bwNearColors = new Set();

  for (let i = 0; i < data.length; i += 4) {
    const r = data[i];
    const g = data[i + 1];
    const b = data[i + 2];
    const a = data[i + 3];

    if (a < 128) {
      transparentPixels++;
      continue;
    }

    const factor = 12;
    const qr = Math.round(r / factor) * factor;
    const qg = Math.round(g / factor) * factor;
    const qb = Math.round(b / factor) * factor;
    const key = `${qr},${qg},${qb}`;

    const isBWNearColor = (r + g + b < 60) || (r + g + b > 710) || (Math.max(r, g, b) - Math.min(r, g, b) < 15);

    if (excludeBW) {
      if (isBWNearColor) {
        excludedBWPixels++;
        continue;
      }
    } else {
      if (isBWNearColor) {
        bwNearColors.add(key);
      }
    }

    if (!colorCounts[key]) {
      colorCounts[key] = { r: qr, g: qg, b: qb, count: 0 };
    }
    colorCounts[key].count++;
    totalValidPixels++;
  }

  const sortedColors = Object.values(colorCounts)
    .sort((a, b) => b.count - a.count);

  if (dev) {
    console.log(`\n=== 颜色统计 ===`);
    console.log(`总像素数: ${info.width * info.height}`);
    console.log(`透明像素: ${transparentPixels}`);
    console.log(`排除的黑白像素: ${excludedBWPixels}`);
    console.log(`有效像素: ${totalValidPixels}`);
    console.log(`量化后的颜色数量: ${sortedColors.length}`);
    console.log(`\n=== 前10个颜色（按出现次数）===`);
    sortedColors.slice(0, 10).forEach((c, i) => {
      const key = `${c.r},${c.g},${c.b}`;
      const isBWNear = bwNearColors.has(key);
      const bwLabel = isBWNear ? ' (黑白临近色)' : '';
      console.log(`${i + 1}. ${rgbToHex(c.r, c.g, c.b)} - 出现 ${c.count} 次 (${Math.round((c.count / totalValidPixels) * 100)}%)${bwLabel}`);
    });
  }

  const finalPalette = [];
  let skippedSimilar = 0;
  
  for (const c of sortedColors) {
    if (finalPalette.length >= maxColors) break;

    const isTooSimilar = finalPalette.some(existing => {
      const dr = existing.rgb.r - c.r;
      const dg = existing.rgb.g - c.g;
      const db = existing.rgb.b - c.b;
      return Math.sqrt(dr*dr + dg*dg + db*db) < 45;
    });

    if (!isTooSimilar) {
      finalPalette.push({
        hex: rgbToHex(c.r, c.g, c.b),
        rgb: { r: c.r, g: c.g, b: c.b },
        count: c.count,
        percentage: Math.round((c.count / totalValidPixels) * 100)
      });
    } else {
      skippedSimilar++;
    }
  }

  if (dev) {
    console.log(`\n=== 最终结果 ===`);
    console.log(`跳过的相似颜色: ${skippedSimilar}`);
    console.log(`最终提取颜色数: ${finalPalette.length}`);
    finalPalette.forEach((c, i) => {
      console.log(`${i + 1}. ${c.hex} - RGB(${c.rgb.r}, ${c.rgb.g}, ${c.rgb.b}) - ${c.percentage}%`);
    });
    console.log('=== 调试结束 ===\n');
  }

  if (finalPalette.length === 0 && sortedColors.length > 0) {
    return sortedColors.slice(0, maxColors).map(c => ({
      hex: rgbToHex(c.r, c.g, c.b),
      rgb: { r: c.r, g: c.g, b: c.b },
      count: c.count,
      percentage: Math.round((c.count / totalValidPixels) * 100)
    }));
  }

  return finalPalette;
};

module.exports = {
  rgbToHex,
  hexToRgb,
  hexToRgba,
  extractPalette
};
